### PR TITLE
ARROW-6205: [C++] ARROW_DEPRECATED warning when including io/interfaces.h

### PR DESCRIPTION
**Pull request by harrism**
ARROW_DEPRECATED on using statements causes warnings when including interfaces.h from CUDA source files (.cu) compiled with nvcc.

This PR works around this issue the same way it is done for MSVC by changing a single preprocessor condition.

Fixes [JIRA 6205 ](https://issues.apache.org/jira/browse/ARROW-6205?jql=project%20%3D%20ARROW%20AND%20resolution%20%3D%20Unresolved%20AND%20text%20~%20%22ARROW_DEPRECATED%22%20ORDER%20BY%20priority%20DESC%2C%20updated%20DESC)

### CODE REVIEW

File: cpp/src/arrow/io/interfaces.h
Comment (emkornfield) [inline cpp/src/arrow/io/interfaces.h]: It seems like the right thing to do might be to delete the using statements altogher based on the TODO above? @kszucs?

Comment (harrism) [reply cpp/src/arrow/io/interfaces.h]: Agreed. I was unsure what "this" referred to in the comment. The whole ifndef and else, or part of it. How long after 0.13 is required? What is it about post-0.13 that makes it possible to remove, etc.

Comment (pitrou) [reply cpp/src/arrow/io/interfaces.h]: Agreed with @emkornfield, we should simply remove those aliases.

Comment (wesm) [reply cpp/src/arrow/io/interfaces.h]: Agreed

Comment (harrism) [reply cpp/src/arrow/io/interfaces.h]: Thanks everyone. Can you explain what from this code block you would like me to remove, specifically? I'm new here. I'm just trying to eliminate potentially build-breaking warnings for RAPIDS libcudf.

https://github.com/apache/arrow/blob/1a3eab50ea49414494d4f7ed91d1e2e3a6f8919f/cpp/src/arrow/io/interfaces.h#L193-L201

Comment (kou) [reply cpp/src/arrow/io/interfaces.h]: These codes are for users who use old Arrow C++. We can remove codes for backward compatibility when we release two or more minor (or major) versions. So we can just apply the following diff:

```diff
diff --git a/cpp/src/arrow/io/interfaces.h b/cpp/src/arrow/io/interfaces.h
index 3a5cfe3d7..678366bd7 100644
--- a/cpp/src/arrow/io/interfaces.h
+++ b/cpp/src/arrow/io/interfaces.h
@@ -190,16 +190,6 @@ class ARROW_EXPORT ReadWriteFileInterface : public RandomAccessFile, public Writ
   ReadWriteFileInterface() { RandomAccessFile::set_mode(FileMode::READWRITE); }
 };
 
-// TODO(kszucs): remove this after 0.13
-#ifndef _MSC_VER
-using WriteableFile ARROW_DEPRECATED("Use WritableFile") = WritableFile;
-using ReadableFileInterface ARROW_DEPRECATED("Use RandomAccessFile") = RandomAccessFile;
-#else
-// MSVC does not like using ARROW_DEPRECATED with using declarations
-using WriteableFile = WritableFile;
-using ReadableFileInterface = RandomAccessFile;
-#endif
-
 }  // namespace io
 }  // namespace arrow
```

Comment (harrism) [reply cpp/src/arrow/io/interfaces.h]: Thanks @kou. Done.

**Review by kou — APPROVED** [review]
+1
We can merge this when CI is green and @harrism confirms that this change resolves the reported warning.

**Review by wesm — APPROVED** [review]
+1

**Comment by harrism** [comment]
@wesm did you intend to close this PR rather than merge it? it appears you agreed with @kou that it should be merged once CI passes.

@kou I can confirm that this change eliminates the warnings when building RAPIDS libcudf.

**Comment by pitrou** [comment]
@harrism It was merged actually, we're just using a dedicated merge script rather than Github's PR merge button.

**Comment by wesm** [comment]
@harrism right, we don't use the GitHub UI to merge PRs, our merge tool creates atomic patches from each pull request and commits those to master, see

https://github.com/apache/arrow/commit/e41ad0d2ccaf96812d902b161d8a0b2b372f1b72

I'm actually a bit surprised a project with many developers like RAPIDS is able to tolerate a "dirty" commit history (https://github.com/rapidsai/cudf/commits/branch-0.9). How do you expect to be able to git-bisect?

**Comment by harrism** [comment]
I've git bisected myself and so have others without problems. log(N) is a very powerful tool.

Thanks for accepting my patch.

**Comment by wesm** [comment]
The number of commits is not an issue -- it's a question of having a mix of "good" (green CI build) and "bad" (failing CI build) in the commit history. In any case, that's a discussion for another time and place :)

**Pull request closed without merging**
